File: Rubyとか/11_SieveOfEratosthenes_CountSemiprimes.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

#include <algorithm>
#include <cmath>
#include <unordered_map>
#include <vector>

vector<int> solution(int N, vector<int> &P, vector<int> &Q) {
  // write your code in C++14 (g++ 6.2.0)

  // 自然数Nを割る最小の素数を列挙する
  int n = N + 1;
  vector<int> prime(n, 0);
  int sqn = (int)sqrt(n);
  for (int i = 2; i <= sqn; ++i) {
    if (prime[i] > 0) continue;
    int k = i * i;
    while (k <= n) {
      if (prime[k] == 0) prime[k] = i;
      k += i;
    }
  }

  // Nまでの半素数を列挙する
  vector<int> semiPrime(n, 0);
  for (int i = 0; i < n; ++i) {
    if (prime[i] == 0) continue;  // 素数を約数に持たない→素数である
    // 約数の中で最小の素数で割る →
    // 割った先の値でもまだ約数を持つならば，その値は半素数ではない(素数2つの積ではない)
    if (prime[i / prime[i]] == 0) semiPrime[i] = 1;
  }

  // 半素数リストの累積和 → 任意の区間内の半素数の数をO(1)で計算できる
  vector<int> cumulative(n, 0);
  for (int i = 1; i < n; ++i) {
    cumulative[i] = cumulative[i - 1] + semiPrime[i];
  }

  // クエリ処理
  int pn = P.size();
  vector<int> ans(pn);
  for (int i = 0; i < pn; ++i) {
    ans[i] = cumulative[Q[i]] - cumulative[P[i]] + semiPrime[P[i]];
  }

  // for(int i = 0; i < prime.size(); ++i) cout << prime[i] << " "; cout <<
  // endl; for(int i = 0; i < semiPrime.size(); ++i) cout << semiPrime[i] << "
  // "; cout << endl; for(int i = 0; i < cumulative.size(); ++i) cout <<
  // cumulative[i] << " "; cout << endl;

  return ans;
}

File: Python/11_SieveOfEratosthenes_CountSemiprimes.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")
import math 

def solution(N, P, Q):
    # write your code in Python 3.6

    # 自然数 a を割れる最小の素数を列挙する(自分自身はカウントしない)
    # → priFactor[a] == 0 ならば a は素数である
    priFactor = [0] * (N + 1)
    for i in range(2, int(math.sqrt(N)) + 1):
        if priFactor[i] != 0: continue  
        ii = i * i
        while ii <= N:
            if priFactor[ii] == 0: priFactor[ii] = i
            ii += i  
    # end of for

    # Nまでの半素数を列挙する
    semiPrimes = [0] * (N + 1)
    for i in range(2, N + 1):
        # 素数を約数に持たない → 素数である
        if priFactor[i] == 0: continue   

        # 自然数 a を割れる最小の素数 p があるとき
        # a / p の値が自身以外の約数を持たないなら半素数である
        # a = p * (a / p) かつ， p， a / p がともに素数なので a は２つの素数の積である
        if priFactor[i // priFactor[i]] == 0:
            semiPrimes[i] = 1
    # end of for 

    # 半素数の有無について累積和をとる
    # → 各クエリに対してO(1)でわかるようになる
    prefixSemiPrimes = [0] * (N + 1) 
    for i in range(1, N + 1):
        prefixSemiPrimes[i] = prefixSemiPrimes[i - 1] + semiPrimes[i]

    m = len(P)
    ans = [0] * m 
    for i in range(m):
        p, q = P[i], Q[i]
        # 累積和ではpのときに増えたかどうかがわからないのでsemiPrimes[p]で判別している
        ans[i] = prefixSemiPrimes[q] - prefixSemiPrimes[p] + semiPrimes[p]
    return ans   
# end of func


            
